### PR TITLE
feat: ✨ add OVHcloud embedding models

### PR DESCRIPTION
**Pull request by philippart-s**
## Change

With this PR I added the OVHcloud embedding models (see https://endpoints.ai.cloud.ovh.net for more details).

## General checklist

- [X] There are no breaking changes
- [X] I have added unit and integration tests for my change
- [X] I have manually run all the unit and integration tests in the module I have added/changed, and they are all green
- [X] I have manually run all the unit and integration tests in the [core](https://github.com/langchain4j/langchain4j/tree/main/langchain4j-core) and [main](https://github.com/langchain4j/langchain4j/tree/main/langchain4j) modules, and they are all green

- [x] I have added/updated the [documentation](https://github.com/langchain4j/langchain4j/tree/main/docs/docs)
- [x] I have added an example in the [examples repo](https://github.com/langchain4j/langchain4j-examples) (only for "big" features)
- [ ] I have added/updated [Spring Boot starter(s)](https://github.com/langchain4j/langchain4j-spring) (if applicable)

## Checklist for adding new model integration

- [x] I have added my new module in the [BOM](https://github.com/langchain4j/langchain4j/blob/main/langchain4j-bom/pom.xml)

## Checklist for adding new embedding store integration

- [x] I have added a `{NameOfIntegration}EmbeddingStoreIT` that extends from either `EmbeddingStoreIT` or `EmbeddingStoreWithFilteringIT`
- [x] I have added my new module in the [BOM](https://github.com/langchain4j/langchain4j/blob/main/langchain4j-bom/pom.xml)

## Checklist for changing existing embedding store integration

- [X] I have manually verified that the `{NameOfIntegration}EmbeddingStore` works correctly with the data persisted using the latest released version of LangChain4j

### CODE REVIEW

**Comment by langchain4j** [comment]
@philippart-s thank you!

**Comment by philippart-s** [comment]
@langchain4j do you need more information to start the review?

**Comment by langchain4j** [comment]
@philippart-s no, just need more hours in the day 😅 I will try to review it this week

**Comment by philippart-s** [comment]
Don't put pressure on yourself it was just to get some info, do this when you can. 
No problem

File: langchain4j-ovh-ai/src/test/java/dev/langchain4j/model/ovhai/OvhAiEmbeddingModelIT.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
+package dev.langchain4j.model.ovhai;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import org.junit.jupiter.api.Test;
+import dev.langchain4j.data.embedding.Embedding;
+import dev.langchain4j.data.segment.TextSegment;
+import dev.langchain4j.model.embedding.EmbeddingModel;
+import dev.langchain4j.model.output.Response;
+
+public class OvhAiEmbeddingModelIT {
```
Comment (langchain4j) [inline langchain4j-ovh-ai/src/test/java/dev/langchain4j/model/ovhai/OvhAiEmbeddingModelIT.java:14]: Please add `@EnabledIfEnvironmentVariable(named = "OVHAI_AI_API_KEY", matches = ".+")`

File: langchain4j-ovh-ai/src/test/java/dev/langchain4j/model/ovhai/OvhAiEmbeddingModelIT.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+        // when
+        Response<Embedding> response = model.embed(textSegment);
+
+        // then
+        assertThat(response.content().vector()).hasSize(768);
+
+        assertThat(response.finishReason()).isNull();
+    }
+}
```
Comment (langchain4j) [inline langchain4j-ovh-ai/src/test/java/dev/langchain4j/model/ovhai/OvhAiEmbeddingModelIT.java:86]: please add a test that embeds multiple segments at once

File: langchain4j-ovh-ai/src/main/java/dev/langchain4j/model/ovhai/OvhAiEmbeddingModel.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    }
+
+    @Override
+    public Response<List<Embedding>> embedAll(List<TextSegment> textSegments) {
+        EmbeddingRequest request = EmbeddingRequest
+            .builder()
+            .input(textSegments.stream().map(TextSegment::text).collect(toList()))
+            .build();
+
+        EmbeddingResponse response = withRetry(() -> client.embed(request), maxRetries);
```
Comment (langchain4j) [inline langchain4j-ovh-ai/src/main/java/dev/langchain4j/model/ovhai/OvhAiEmbeddingModel.java]: Do you know if there are any limits on how many segments can be embedded in one API call and how big each segment can be?

**Review by ghost — COMMENTED** [review]
@philippart-s thanks a lot!

File: langchain4j-ovh-ai/src/test/java/dev/langchain4j/model/ovhai/OvhAiEmbeddingModelIT.java (hunk 1 above)
Comment (philippart-s) [reply langchain4j-ovh-ai/src/test/java/dev/langchain4j/model/ovhai/OvhAiEmbeddingModelIT.java:14]: fixed with https://github.com/langchain4j/langchain4j/pull/1355/commits/85fed3ea81f3346927adb52c3ed7968e66d93875

File: langchain4j-ovh-ai/src/test/java/dev/langchain4j/model/ovhai/OvhAiEmbeddingModelIT.java (hunk 2 above)
Comment (philippart-s) [reply langchain4j-ovh-ai/src/test/java/dev/langchain4j/model/ovhai/OvhAiEmbeddingModelIT.java:86]: I have one problem: our embedding endpoints don't support the multiple segments at once 😭.

I have three options:

1. This is a limitation and I throw an exception if there are more than one segment
2. I'll develop the multiple call in the client (not very good in performance but allows the presence of the multiple segments feature)
3. I'll wait for my developers develop this feature in our endpoints and after I update the PR

I can do 2 then open a new PR with 3 if you want.
Or we wait 3 if you prefer, tell me 😉 .

File: langchain4j-ovh-ai/src/main/java/dev/langchain4j/model/ovhai/OvhAiEmbeddingModel.java (hunk 3 above)
Comment (philippart-s) [reply langchain4j-ovh-ai/src/main/java/dev/langchain4j/model/ovhai/OvhAiEmbeddingModel.java]: As mentioned (https://github.com/langchain4j/langchain4j/pull/1355#discussion_r1675860105), this feature is not yet developed. When it is, where do you want me to mention it?

File: langchain4j-ovh-ai/src/test/java/dev/langchain4j/model/ovhai/OvhAiEmbeddingModelIT.java (hunk 2 above)
Comment (langchain4j) [reply langchain4j-ovh-ai/src/test/java/dev/langchain4j/model/ovhai/OvhAiEmbeddingModelIT.java:86]: "2 now, 3 later" sounds good :)

File: langchain4j-ovh-ai/src/main/java/dev/langchain4j/model/ovhai/OvhAiEmbeddingModel.java (hunk 3 above)
Comment (langchain4j) [reply langchain4j-ovh-ai/src/main/java/dev/langchain4j/model/ovhai/OvhAiEmbeddingModel.java]: ok, what about "how big each segment can be?"

Comment (philippart-s) [reply langchain4j-ovh-ai/src/main/java/dev/langchain4j/model/ovhai/OvhAiEmbeddingModel.java]: Today we support this:
    - input: 512 tokens
    - output: embedding size is 768

File: langchain4j-ovh-ai/pom.xml [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+            <scope>test</scope>
+        </dependency>
+
+    </dependencies>
+
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-failsafe-plugin</artifactId>
```
Comment (langchain4j) [inline langchain4j-ovh-ai/pom.xml]: this plugin can be removed along with `skipOvhAIITs`

File: langchain4j-ovh-ai/src/test/java/dev/langchain4j/model/ovhai/OvhAiEmbeddingModelIT.java (hunk 2 above)
Comment (philippart-s) [reply langchain4j-ovh-ai/src/test/java/dev/langchain4j/model/ovhai/OvhAiEmbeddingModelIT.java:86]: @langchain4j the commit https://github.com/langchain4j/langchain4j/pull/1355/commits/771037f2ac79b3aa50e54ac6b847df2bbb6ee044 is a first version to a _embedAll_ solution with its corresponding test.

Not sure it's the more efficient but it is a first implementation 😉

**Review by ghost — COMMENTED** [review]
@philippart-s please add documentation page under `docs/docs/integrations/embedding-models`

File: langchain4j-ovh-ai/pom.xml (hunk 4 above)
Comment (philippart-s) [reply langchain4j-ovh-ai/pom.xml]: done with https://github.com/langchain4j/langchain4j/pull/1355/commits/655bcf0ab1ba596dc0177a05e311c19306afab1d

**Comment by philippart-s** [comment]
> @philippart-s please add documentation page under `docs/docs/integrations/embedding-models`

done with https://github.com/langchain4j/langchain4j/pull/1355/commits/a1af1564ef68ce92507bc6be2ca9aa3a501479f9

@langchain4j I also created a PR (https://github.com/langchain4j/langchain4j-examples/pull/96) in the https://github.com/langchain4j/langchain4j-examples repository. I don't know how you want synchronise these two PRs 😉

**Review by ghost — APPROVED** [review]
@philippart-s thanks a lot!

**Comment by langchain4j** [comment]
@philippart-s I will merge the example PR as soon as I release a new version with this new module (hopefully next week)

**Comment by philippart-s** [comment]
> @philippart-s I will merge the example PR as soon as I release a new version with this new module (hopefully next week)

Yes 🤘, thank you very much, in the next few weeks I'll be adding more features like the classic ITM templates 😄.

**Comment by langchain4j** [comment]
Hi @philippart-s, would it be possible to provide an API key with free credits to test this integration on CI? It was a pain to update the free key every week and now I can't create the new key, the process does not seem straightforward. Thank you! 🙏

**Comment by philippart-s** [comment]
Yes we have just changed how to generate the token (there is no impact on the code, only the way to generate it).

I will check internally with my core team how we can provide a more stable API Key for the CI 😉
How is this token stored and by whom is it known?
And do you have an estimate of the number of requests per minute/month that are made?

**Comment by langchain4j** [comment]
@philippart-s thanks a lot, that would be awesome!
The token is stored in GitHub secrets and injected via environment variable when workflow with integration tests is running.
Only I have access to it.
The tests are running every night 2 times. As far as I see from `OvhAiEmbeddingModelIT`, 3 requests are done. So 6 requests per day in total. We can add sleep between tests if required.

**Comment by philippart-s** [comment]
@langchain4j how can I set the new token?

**Comment by langchain4j** [comment]
@philippart-s could you please send it to me on langchain4j@gmail.com and I will set it into GitHub secrets

**Comment by dliubarskyi** [comment]
Hi @philippart-s, `OvhAiEmbeddingModelIT` stopped working this night.

I am getting this error:
```
java.lang.RuntimeException: status code: 404; body: {
  "message":"no Route matched with those values",
  "request_id":"aea5ab03f9750d707f40b14d52ba5b5a"
}
```

From your website it seems that only OpenAI API is supported? Should we deprecate/remove this module?

**Comment by philippart-s** [comment]
Hi @dliubarskyi ,

I will have a look and get back to you ASAP.

**Comment by philippart-s** [comment]
@dliubarskyi, the error is due to the bge-m3 model being decommissioned. I could replace it in the code, but I think your remark is valid.

To remove some non-essential code, since AI Endpoints are fully compatible with OpenAI, we could reconsider having dedicated code for AI Endpoints.

Do you think it would be possible to keep the documentation (I can update it) to explain how to use AI Endpoints with LangChain4j, even if you remove the code part?

**Comment by dliubarskyi** [comment]
@philippart-s I don't mind keeping the dedicated `OvhAiEmbeddingModel` (even though it can be replaced by `OpenAiEmbeddingModel`), but tests need to work. Can you provide a working `baseUrl` for the tests?

**Comment by philippart-s** [comment]
ok, I'll try to update this ASAP with the new URL and correct model.
I'm not sure to have time before this week-end to do the modification.

**Comment by dliubarskyi** [comment]
@philippart-s sure, no rush, but please don't forget about it

**Comment by dliubarskyi** [comment]
@philippart-s friendly ping

**Comment by philippart-s** [comment]
Yes, I will do that ASAP. I'm at Devoxx France, I try to do the update before the end of the week.

**Comment by philippart-s** [comment]
@dliubarskyi I took the time to analyze the issue. Since we are now fully compatible with OpenAI embedding model requests, using the new AI Endpoints URL would require rewriting a significant part of the code to align with the OpenAI client.

I’m not sure how you’d like to proceed:
1. Remove the OVH-AI provider
2. Update the tests and documentation to use the OpenAI provider with OVHcloud models, and remove the internal OVHcloud-specific code
3. Keep only the documentation explaining how to use OVHcloud embedding models via the OpenAI provider
4. Other?

**Comment by dliubarskyi** [comment]
@philippart-s just to make sure I understand correctly: OVHcloud deprecated their specific API for embeddings and only have OpenAI-compatible embeddings API left, right?

**Comment by philippart-s** [comment]
Yes 😊

**Comment by dliubarskyi** [comment]
Then this means that `OvhAiEmbeddingModel` and the whole `langchain4j-ovh-ai` module cannot be used at the moment (at all), right? Then it would make sense to deprecate `OvhAiEmbeddingModel` and `langchain4j-ovh-ai` maven module and update OVH documentation to show how to use `langchain4j-open-ai` module instead. WDYT?

**Comment by philippart-s** [comment]
Yes I agree with you, I see how to update the documentation part but I need to search how to make a maven module deprecated.

I will start a new PR to do these updatesif it's ok for you.

**Comment by dliubarskyi** [comment]
@philippart-s I think deprecating every public class in the module should be enough. And documentation. Then after a few releases we will just remove it.

```java
/**
 @deprecated use {@code OpenAiEmbeddingModel} from {@code langchain4j-open-ai} module instead
 */
@Deprecated(forRemoval = true, since = "1.14.0")
```

**Comment by philippart-s** [comment]
@dliubarskyi PR done here: https://github.com/langchain4j/langchain4j/pull/5016/

Do you think I could do the same kind of documentation for other models from OVH that are compatible with OpenAI?

**Comment by dliubarskyi** [comment]
> Do you think I could do the same kind of documentation for other models from OVH that are compatible with OpenAI?

It's up to you!

**Pull request merged**